consensus/state_processing/src/per_epoch_processing/single_pass.rs

```diff
@@ -0,0 +1,628 @@
+
+impl Default for SinglePassConfig {
+    fn default() -> SinglePassConfig {
+        Self::enable_all()
+    }
+}
+
+impl SinglePassConfig {
+    pub fn enable_all() -> SinglePassConfig {
+        Self {
+            inactivity_updates: true,
+            rewards_and_penalties: true,
+            registry_updates: true,
+            slashings: true,
+            effective_balance_updates: true,
+        }
+    }
+
+    pub fn disable_all() -> SinglePassConfig {
+        SinglePassConfig {
+            inactivity_updates: false,
+            rewards_and_penalties: false,
+            registry_updates: false,
+            slashings: false,
+            effective_balance_updates: false,
+        }
+    }
+}
+
+/// Values from the state that are immutable throughout epoch processing.
+struct StateContext {
+    current_epoch: Epoch,
+    next_epoch: Epoch,
+    is_in_inactivity_leak: bool,
+    total_active_balance: u64,
+    churn_limit: u64,
+    fork_name: ForkName,
+}
+
+struct RewardsAndPenaltiesContext {
+    unslashed_participating_increments_array: [u64; NUM_FLAG_INDICES],
+    active_increments: u64,
+}
+
+struct SlashingsContext {
+    adjusted_total_slashing_balance: u64,
+    target_withdrawable_epoch: Epoch,
+}
+
+struct EffectiveBalancesContext {
+    downward_threshold: u64,
+    upward_threshold: u64,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ValidatorInfo {
+    pub index: usize,
+    pub effective_balance: u64,
+    pub base_reward: u64,
+    pub is_eligible: bool,
+    pub is_slashed: bool,
+    pub is_active_current_epoch: bool,
+    pub is_active_previous_epoch: bool,
+    // Used for determining rewards.
+    pub previous_epoch_participation: ParticipationFlags,
+    // Used for updating the progressive balances cache for next epoch.
+    pub current_epoch_participation: ParticipationFlags,
+}
+
+impl ValidatorInfo {
+    #[inline]
+    pub fn is_unslashed_participating_index(&self, flag_index: usize) -> Result<bool, Error> {
+        Ok(self.is_active_previous_epoch
+            && !self.is_slashed
+            && self
+                .previous_epoch_participation
+                .has_flag(flag_index)
+                .map_err(|_| Error::InvalidFlagIndex(flag_index))?)
+    }
+}
+
+pub fn process_epoch_single_pass<E: EthSpec>(
+    state: &mut BeaconState<E>,
+    spec: &ChainSpec,
+    conf: SinglePassConfig,
+) -> Result<ParticipationEpochSummary<E>, Error> {
+    initialize_epoch_cache(state, spec)?;
+    initialize_progressive_balances_cache(state, spec)?;
+    state.build_exit_cache(spec)?;
+
+    let previous_epoch = state.previous_epoch();
+    let current_epoch = state.current_epoch();
+    let next_epoch = state.next_epoch()?;
+    let is_in_inactivity_leak = state.is_in_inactivity_leak(previous_epoch, spec)?;
+    let total_active_balance = state.get_total_active_balance()?;
+    let churn_limit = state.get_churn_limit(spec)?;
+    let activation_churn_limit = state.get_activation_churn_limit(spec)?;
```

Comment:
The functions `get_total_active_balance()`, `get_churn_limit()`, and `get_activation_churn_limit()` all require the **committee** caches to be built but I don't see anything that ensures this?

Reply:
Ah I see they are built in `process_epoch` but not necessarily in all the `_slow()` functions which call `process_epoch_single_pass`

Reply:
good catch! addressed in https://github.com/sigp/lighthouse/pull/5386